modules/mauth-common/src/main/java/com/mdsol/mauth/util/MAuthSignatureHelper.java

```diff
@@ -224,35 +223,99 @@ public static String generateEncryptedQueryParams(String query) throws IOExcepti
         if (encryptedQueryParams.length() > 0) {
           encryptedQueryParams.append('&');
         }
-        encryptedQueryParams.append(urlEncodeValue(name)).append('=').append(urlEncodeValue(value));
+        encryptedQueryParams.append(encordQueryString(name)).append('=').append(encordQueryString(value));
       }
     }
 
     return encryptedQueryParams.toString();
   }
 
   /**
-   * encode a string value using `UTF-8` encoding scheme
+   * decode and sort a encoded query string
+   * @param encodedQuery a url-encoded string
+   * @return sorted decoded query parameters
+   */
+  private static String[] sortQueryParameters(String encodedQuery) {
+    String[] params = encodedQuery.split("&");
+    List<String> list = new ArrayList<String>();
+
+    for (String param : params) {
+      String [] keyPair = param.split("=");
+      if (keyPair.length > 0) {
+        String name = param.split("=")[0];
+        String value = keyPair.length > 1 ? param.split("=")[1] : "";
+        list.add(dencordQueryString(name) + "=" + dencordQueryString(value));
+      }
+    }
+    String[] decodedParams = list.toArray(new String[0]);
+    Arrays.sort(decodedParams);
+
+    return decodedParams;
+  }
+
+  /**
+   * encode a query string using `UTF-8` encoding scheme
    * @param value
    * @return encoded String
-   *
-   * See https://stackoverflow.com/questions/10786042/java-url-encoding-of-query-string-parameters
    */
-  private static String urlEncodeValue(String value) {
+  private static String encordQueryString(String value) {
     if (value == null ||  value.isEmpty())
       return value;
 
     try {
       String encodedValue = URLEncoder.encode(value, StandardCharsets.UTF_8.toString());
-      encodedValue = encodedValue.replace("+", "%20");
-      encodedValue = encodedValue.replace("%7E", "~");
-      encodedValue = encodedValue.replace("*", "%2A");
+      encodedValue = encodedValue.replace("+", "%20").replace("%7E", "~").replace("*", "%2A");
```

Comment:
In the "dencordQueryString" method, `+` is already replaced with `(space)`:

https://github.com/mdsol/mauth-jvm-clients/blob/a405eedbe7a966f49fca969b1dbff12240bca5ff/modules/mauth-common/src/main/java/com/mdsol/mauth/util/MAuthSignatureHelper.java#L285

so I think you don't need to replace it here but it doesn't hurt so maybe better to keep it 😅

Reply:
agree